File: 07_valid_parameters1.ts
// optimized
var isValid1 = function(s: string): boolean {
    const stack = []

    for (let i = 0; i < s.length; i++) {
        const cur = s[i]
        console.log('cur: ', cur)

        if (stack.length) {
            const last = stack[stack.length - 1]

            if (isPair(last, cur)) {
                stack.pop()
                continue
            }
        }
        stack.push(cur)
    }

    return stack.length === 0
};

var isPair = function(last: string, cur: string): boolean {
    return (
        (last === '(' && cur === ')') ||
        (last === '{' && cur === '}') ||
        (last === '[' && cur === ']')
    )
};

const result_08 = isValid1("[()]")  // true
// const result_08 = isValid1("[")     // false
// const result_08 = isValid1("[[")    // false
// const result_08 = isValid1("([]){") // false
// const result_08 = isValid1("([)]")     // false
console.log("result: ", result_08)

File: 07_valid_parameters1.rs
struct Solution {}

// optimized
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();


        for i in s.chars() {
            match i {
                '{' => stack.push('}'),
                '(' => stack.push(')'),
                '[' => stack.push(']'),
                '}' | ')' | ']' if Some(i) != stack.pop() => return false, // invalidation, terminate the loop and the program
                _ => (),
            }
        }

        // case: "["
        return stack.is_empty()
    }
}

// [()]
fn main() {
    // let result: bool = Solution::is_valid(("[()]").to_string());      // true
    // let result: bool = Solution::is_valid("[".to_string());           // false
    // let result: bool = Solution::is_valid("[[".to_string());          // false
    // let result: bool = Solution::is_valid("([]){".to_string());       // false
    let result: bool = Solution::is_valid("(())]".to_string());        // false
    println!("{}", result);
}
